File: 27.cpp
class Solution {
public:
    int removeElement(vector<int>& nums, int val) {
        if(nums.empty())  return 0;
        int lastNot=nums.size()-1;
        while(nums[lastNot]==val && lastNot>=0)  lastNot--;
        
        for (int i=0; i<nums.size() && i<lastNot ; i++){
            if(nums[i] == val){
                swap(nums[i], nums[lastNot--]);
                while(nums[lastNot] == val && lastNot>=0){
                    lastNot--;
                }
            }
        }
        return lastNot+1;
    }
};

File: golang/27/27.go
package main

func removeElement(nums []int, val int) int {
	if len(nums) == 0 {
		return 0
	}

	index := len(nums) - 1
	for nums[index] == val {
		index--
		if index < 0 {
			return 0
		}
	}

	for i := 0; i < len(nums) && i < index; i++ {
		if nums[i] == val {
			for index > i {
				if nums[index] != val {
					nums[i], nums[index] = nums[index], nums[i]
					index--
					break
				}
				index--
			}
		}
	}
	if nums[index] == val {
		index--
	}
	return index + 1
}

